cassovary-core/src/main/scala/com/twitter/cassovary/algorithms/linkanalysis/PageRank.scala

```diff
@@ -44,15 +44,28 @@ case class PageRankIterationState(pageRank: Array[Double], error: Double, iterat
 class PageRank(graph: DirectedGraph[Node], params: PageRankParams = PageRankParams())
   extends LinkAnalysis[PageRankIterationState](graph, params, "pagerank") {
 
-  private val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private lazy val outboundCount = new Array[Double](graph.maxNodeId + 1)
+  private val nodeCount = graph.nodeCount
 
-  graph foreach { node => efficientNeighbors(node) foreach { nbr => outboundCount(nbr) += 1} }
+  private val dangling = if (isInStored) {
+    val nonDangling = graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      val neighbors = efficientNeighbors(node)
+      neighbors foreach { nbr => outboundCount(nbr) += 1 }
+      partialSet ++ neighbors.toSet
+    }
+    graph.map { _.id }.toSet -- nonDangling
+  }
+  else {
+    graph.foldLeft(Set.empty[Int]){ (partialSet, node) =>
+      if (efficientNeighbors(node).isEmpty) partialSet + node.id else partialSet
+    }
+  }
```

Comment:
So that is why I check isInStored here.  Basically, I don't need to build the outboundCount array if the graph is out stored or both stored since that attribute is stored at the node level.  I only need to build that array if it is in stored (which is an expensive calculation because we need to visit each node).  By checking isInStored here, we can save on the calculation for outboundNodes in the determination of dangling nodes.